File: expense-report/typescript/tests/ReportBuilder.ts
import type { ExpenseItem } from '../src/ExpenseItem.js';
import { Report } from '../src/Report.js';
import { ExpenseItemBuilder } from './ExpenseItemBuilder.js';

export class ReportBuilder {
  private _employeeName = 'Alice Johnson';
  private _expenses: ExpenseItem[] = [];
  private _submit = false;
  private _approve = false;
  private _rejectReason: string | null = null;

  withEmployeeName(name: string): this { this._employeeName = name; return this; }

  withExpense(item: ExpenseItem): this { this._expenses.push(item); return this; }

  withExpenseFrom(configure: (b: ExpenseItemBuilder) => void): this {
    const builder = new ExpenseItemBuilder();
    configure(builder);
    this._expenses.push(builder.build());
    return this;
  }

  submitted(): this { this._submit = true; return this; }

  approved(): this { this._submit = true; this._approve = true; return this; }

  rejected(reason = 'Policy violation'): this {
    this._submit = true;
    this._rejectReason = reason;
    return this;
  }

  build(): Report {
    const report = new Report(this._employeeName);
    for (const expense of this._expenses) {
      report.addExpense(expense);
    }
    if (this._submit) report.submit();
    if (this._approve) report.approve();
    if (this._rejectReason !== null) report.reject(this._rejectReason);
    return report;
  }
}

File: expense-report/python/tests/report_builder.py
from __future__ import annotations

from typing import Callable

from expense_report import ExpenseItem, Report

from .expense_item_builder import ExpenseItemBuilder


class ReportBuilder:
    def __init__(self) -> None:
        self._employee_name = "Alice Johnson"
        self._expenses: list[ExpenseItem] = []
        self._submit = False
        self._approve = False
        self._reject_reason: str | None = None

    def with_employee_name(self, name: str) -> ReportBuilder:
        self._employee_name = name
        return self

    def with_expense(self, item: ExpenseItem) -> ReportBuilder:
        self._expenses.append(item)
        return self

    def with_expense_from(self, configure: Callable[[ExpenseItemBuilder], None]) -> ReportBuilder:
        builder = ExpenseItemBuilder()
        configure(builder)
        self._expenses.append(builder.build())
        return self

    def submitted(self) -> ReportBuilder:
        self._submit = True
        return self

    def approved(self) -> ReportBuilder:
        self._submit = True
        self._approve = True
        return self

    def rejected(self, reason: str = "Policy violation") -> ReportBuilder:
        self._submit = True
        self._reject_reason = reason
        return self

    def build(self) -> Report:
        report = Report(self._employee_name)
        for expense in self._expenses:
            report.add_expense(expense)
        if self._submit:
            report.submit()
        if self._approve:
            report.approve()
        if self._reject_reason is not None:
            report.reject(self._reject_reason)
        return report
